### PR TITLE
[test] Apply maximal flags to tx_valid tests and minimal flags to tx_invalid tests

### PR DESCRIPTION
**Pull request by jl2012**
The first commit inverts the meaning of verifyFlags for tx_valid tests, as flags being excluded. All flags are applied by default, except those found in verifyFlags. This makes sure that a new or existing flag won't invalidate a tx by accident.

The second commit reduces the number of validation flags used for tx_invalid tests, to a minimally required set to fail a test. This makes sure that a tx failed due to the tested flags, not unexpected effects of some other flags. It also uses "BADTX" to indicate tests not passing CheckTransaction(), vs. those failing script execution.

(If a test is expected to fail due to multiple independent flags, multiple tests should be used)

The third commit verifies that the flags excluded in tx_valid and included in tx_invalid are indeed the minimal set. In tx_valid, it adds back the excluded flags individually and expects it to fail. In tx_invalid, it removes the included flags individually and expects it to pass.

This process helped me to identify and fix some buggy tests:
- Remove unnecessary OP_1 at the end of most OP_CLTV and OP_CSV tx_valid tests, so there is no need to exclude CLEANSTACK
- An OP_CSV tx_valid test missed an OP_ADD, and is added back
- 2 witness tests were found with empty vout, so they failed due to CheckTransaction(), not script tests. Corrected by filling in proper vout.

### CODE REVIEW

File: src/test/data/tx_invalid.json [hunk 1]
```diff
@@ -82,174 +82,174 @@
...
-"010000000100010000000000000000000000000000000000000000000000000000000000000000000004005194b1010000000100000000000000000002000000", "P2SH,CHECKLOCKTIMEVERIFY"],
+"010000000100010000000000000000000000000000000000000000000000000000000000000000000004005194b1010000000100000000000000000002000000", "CHECKLOCKTIMEVERIFY"],
 
 ["Input locked"],
 [[["0000000000000000000000000000000000000000000000000000000000000100", 0, "0 CHECKLOCKTIMEVERIFY 1"]],
-"010000000100010000000000000000000000000000000000000000000000000000000000000000000000ffffffff0100000000000000000000000000", "P2SH,CHECKLOCKTIMEVERIFY"],
+"010000000100010000000000000000000000000000000000000000000000000000000000000000000000ffffffff0100000000000000000000000000", "CHECKLOCKTIMEVERIFY"],
 [[["0000000000000000000000000000000000000000000000000000000000000100", 0, "0"]],
-"01000000010001000000000000000000000000000000000000000000000000000000000000000000000251b1ffffffff0100000000000000000002000000", "P2SH,CHECKLOCKTIMEVERIFY"],
+"01000000010001000000000000000000000000000000000000000000000000000000000000000000000251b1ffffffff0100000000000000000002000000", "NONE"],
```
Comment (jl2012) [inline src/test/data/tx_invalid.json]: I think there is something wrong with this test, but I am not sure about its intention

File: src/test/data/tx_invalid.json [hunk 2]
```diff
@@ -82,174 +82,174 @@
...
+"010000000200010000000000000000000000000000000000000000000000000000000000000000000000ffffffff00020000000000000000000000000000000000000000000000000000000000000100000000000000000100000000000000000000000000", "CHECKLOCKTIMEVERIFY"],
 
 ["Argument/tx height/time mismatch, both versions"],
 [[["0000000000000000000000000000000000000000000000000000000000000100", 0, "0 CHECKLOCKTIMEVERIFY 1"]],
-"01000000010001000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000065cd1d", "P2SH,CHECKLOCKTIMEVERIFY"],
+"01000000010001000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000000065cd1d", "CHECKLOCKTIMEVERIFY"],
 [[["0000000000000000000000000000000000000000000000000000000000000100", 0, "0"]],
-"01000000010001000000000000000000000000000000000000000000000000000000000000000000000251b100000000010000000000000000000065cd1d", "P2SH,CHECKLOCKTIMEVERIFY"],
+"01000000010001000000000000000000000000000000000000000000000000000000000000000000000251b100000000010000000000000000000065cd1d", "NONE"],
```
Comment (jl2012) [inline src/test/data/tx_invalid.json]: Another problematic test with unclear intention

File: src/test/transaction_tests.cpp [hunk 3]
```diff
@@ -238,28 +267,27 @@ BOOST_AUTO_TEST_CASE(tx_invalid)
...
-                {
-                    BOOST_ERROR("Bad test: " << strTest);
-                    break;
+            std::vector<unsigned int> verify_flags = ParseScriptFlags(test[2].get_str());
+            if (!CheckTx(tx, mapprevOutScriptPubKeys, mapprevOutValues, verify_flags[0], txdata, strTest, false))
+                BOOST_ERROR("Tx unexpectedly passed: " << strTest);
+            for (size_t f = 1; f < verify_flags.size(); f++) {
+                unsigned int flags = verify_flags[f];
+                if (!(flags & SCRIPT_VERIFY_CLEANSTACK)) {
+                    flags &= ~SCRIPT_VERIFY_WITNESS;
```
Comment (practicalswift) [inline src/test/transaction_tests.cpp]: Should the implicit conversion from negative to unsigned be made explicit here?

Comment (jl2012) [reply src/test/transaction_tests.cpp]: Done. Also fixed a bug as I swapped SCRIPT_VERIFY_CLEANSTACK and SCRIPT_VERIFY_WITNESS here

**Comment by jl2012** [comment]
Added a commit to verify all validation flags are backward compatible (softfork), like #10699

File: src/test/script_tests.cpp [hunk 4]
```diff
@@ -155,37 +159,59 @@ CMutableTransaction BuildSpendingTransaction(const CScript& scriptSig, const CSc
...
+        }
     }
 
 #if defined(HAVE_CONSENSUS_LIB)
     CDataStream stream(SER_NETWORK, PROTOCOL_VERSION);
     stream << tx2;
-    int libconsensus_flags = flags & bitcoinconsensus_SCRIPT_FLAGS_VERIFY_ALL;
-    if (libconsensus_flags == flags) {
+    int libconsensus_flags = main_flags & bitcoinconsensus_SCRIPT_FLAGS_VERIFY_ALL;
+    if (expect || libconsensus_flags == (int)main_flags) {
```
Comment (jl2012) [inline src/test/script_tests.cpp]: rebased and fixed a bug here. Without `expect ||`, all valid script tests were skipped due to the use of standardness flags. If a script is valid with standardness flags, removing these flags shouldn't make it invalid.

This also fixed an existing bug that many tests containing standardness flags (mostly STRICTENC) are unnecessarily skipped.

**Comment by practicalswift** [comment]
@jl2012 Could you rebase this PR on master and push to have Travis test the rebased version? I think there might be some UBSan runtime errors that needs to be addressed.

**Comment by MarcoFalke** [comment]
No activity for a year. Closing for now, let me know when you want to work on this again

**Comment by MarcoFalke** [comment]
(last commit hasn't been picked up yet)

**Pull request closed without merging**